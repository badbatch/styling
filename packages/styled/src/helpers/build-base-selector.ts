import { kebabCase } from "lodash";
import shortid from "shortid";

export default function buildBaseSelector(componentName: string, prefix?: string) {
  return `${prefix ? `${prefix}-` : ""}${kebabCase(componentName)}-${shortid.generate()}`;
}
