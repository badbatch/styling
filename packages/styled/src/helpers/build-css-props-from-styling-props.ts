import { PlainObject } from "@styling/types";
import { kebabCase } from "lodash";
import { StylingCSSVariables } from "../types";

export default function buildCSSPropsFromStylingProps(combo: string[], cssVariableProps: StylingCSSVariables) {
  const props: PlainObject = {};

  cssVariableProps.forEach(([key, value]) => {
    props[key] = `var(--${kebabCase(key)}${value ? `, ${value}` : ""})`;
  });

  combo.forEach(prop => {
    const split = prop.split("::");

    if (split.length === 2) {
      props[split[0]] = split[1];
    } else {
      props[split[0]] = true;
    }
  });

  return props;
}
