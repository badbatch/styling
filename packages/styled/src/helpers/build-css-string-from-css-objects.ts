import autoprefixer from "autoprefixer";
import { renderSync } from "node-sass";
import postcss from "postcss";
import postcssJs from "postcss-js";
import { StringDecoder } from "string_decoder";
import { PropKeyComboCSS } from "../types";

export default function buildCSSStringFromCSSObjects(propKeyComboCSS: PropKeyComboCSS) {
  return propKeyComboCSS.reduce((css, entry) => {
    const result = postcss([autoprefixer]).process(
      { [`.${entry[1].selector}`]: entry[1].css },
      {
        from: undefined,
        parser: postcssJs.parse,
      },
    );

    css += `${new StringDecoder("utf8").write(renderSync({ data: result.css }).css)}\n`;
    return css;
  }, "");
}
