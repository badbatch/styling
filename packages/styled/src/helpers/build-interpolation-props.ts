import { CSSVariablePropList, PlainObject } from "@styling/types";
import { kebabCase } from "lodash";

export default function buildInterpolationProps(propKeyCombo: string[], cssVariablePropList: CSSVariablePropList) {
  const props: PlainObject = {};

  /**
   * TODO: Need to implement some sort of fallback
   * for ie11 support, maybe using new config option
   * or babel options to detect ie11 support and if
   * required hardcode the fallback value.
   */

  cssVariablePropList.forEach(([key, value]) => {
    props[key] = `var(--${kebabCase(key)}${value ? `, ${value}` : ""})`;
  });

  propKeyCombo.forEach(prop => {
    const split = prop.split("::");

    if (split.length === 2) {
      props[split[0]] = split[1];
    } else {
      props[split[0]] = true;
    }
  });

  return props;
}
