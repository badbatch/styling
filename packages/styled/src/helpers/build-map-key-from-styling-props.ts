import { kebabCase } from "lodash";

export default function buildMapKeyFromStylingProps(propNameCombo: string[]) {
  return propNameCombo
    .map(propName => kebabCase(propName))
    .sort()
    .join("--");
}
