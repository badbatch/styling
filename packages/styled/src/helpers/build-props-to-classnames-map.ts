import { Interpolation, StringObject } from "@styling/types";
import { existsSync, removeSync } from "fs-extra";
import { resolve } from "path";
import { STYLING_CSS_FILENAME } from "../constants";
import { Metadata, SelectorCSS, StylingCSSVariables } from "../types";
import buildBaseSelector from "./build-base-selector";
import buildCSSPropsFromStylingProps from "./build-css-props-from-styling-props";
import buildCSSStringFromCSSObjects from "./build-css-string-from-css-objects";
import buildMapKeyFromStylingProps from "./build-map-key-from-styling-props";
import buildSelectorFromStylingProps from "./build-selector-from-styling-props";
import collateCSS from "./collate-css";
import dedupeCSS from "./dedupe-css";
import loadStylingConfig from "./load-styling-config";
import { error, info } from "./log";
import writeCSS from "./write-css";

export default function buildPropsToClassNamesMap(
  propNameCombos: string[][],
  cssVariableProps: StylingCSSVariables,
  interpolations: Interpolation[],
  { componentName, sourceDir }: Metadata,
): StringObject {
  const { outputPath, selectorPrefix, theme } = loadStylingConfig({ componentName, sourceDir });
  const fullOutputPath = resolve(outputPath, STYLING_CSS_FILENAME);

  if (existsSync(fullOutputPath)) {
    removeSync(fullOutputPath);
  }

  const baseSelector = buildBaseSelector(componentName, selectorPrefix);
  const baseCSS = collateCSS(interpolations, buildCSSPropsFromStylingProps([], cssVariableProps), theme);

  const selectorCSS: SelectorCSS = {};

  if (Object.keys(baseCSS).length) {
    selectorCSS[baseSelector] = {
      css: baseCSS,
      key: "",
    };
  }

  propNameCombos.forEach(combo => {
    const comboCSS = dedupeCSS(
      collateCSS(interpolations, buildCSSPropsFromStylingProps(combo, cssVariableProps), theme),
      selectorCSS[baseSelector]?.css,
    );

    if (Object.keys(comboCSS).length) {
      selectorCSS[buildSelectorFromStylingProps(baseSelector, combo)] = {
        css: comboCSS,
        key: buildMapKeyFromStylingProps(combo),
      };
    }
  });

  info("Generating css from css objects");
  const css = buildCSSStringFromCSSObjects(selectorCSS);

  info(`Writing css to ${fullOutputPath}`, css);

  try {
    writeCSS(css, fullOutputPath);
  } catch (e) {
    error("Writing css failed", e);
  }

  return Object.keys(selectorCSS).reduce((map: StringObject, selector) => {
    const { key } = selectorCSS[selector];
    if (key) map[key] = selector;
    return map;
  }, {});
}
