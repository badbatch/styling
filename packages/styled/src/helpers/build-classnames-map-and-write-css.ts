import { error, getFullOutputPath, info, loadStylingConfig, verbose } from "@styling/helpers";
import { CSSVariablePropList, Interpolation, Metadata, StringObject } from "@styling/types";
import { existsSync, readFileSync } from "fs-extra";
import { CSS_FILE_EXT } from "../constants";
import buildCSSObjects from "./build-css-objects";
import buildCSSStringFromCSSObjects from "./build-css-string-from-css-objects";
import mergeNewCSSIntoExisting from "./merge-new-css-into-existing";
import writeCSS from "./write-css";

export default function buildClassNamesMapAndWriteCSS(
  propKeyCombos: string[][],
  cssVariablePropList: CSSVariablePropList,
  interpolations: Interpolation[],
  { componentName, propList, sourceFilename }: Metadata,
) {
  const config = loadStylingConfig({ componentName, sourceFilename });

  info("Building css objects");

  const propKeyComboCSS = buildCSSObjects(
    propKeyCombos,
    cssVariablePropList,
    interpolations,
    { componentName, propList, sourceFilename },
    config,
  );

  const { cssOutputPath } = config;

  info("Generating css from css objects");
  let css = buildCSSStringFromCSSObjects(propKeyComboCSS);
  const fullOutputPath = getFullOutputPath(cssOutputPath, sourceFilename, { exclude: "src", extension: CSS_FILE_EXT });

  if (existsSync(fullOutputPath)) {
    info("Merging new css into existing css");
    css = mergeNewCSSIntoExisting(css, readFileSync(fullOutputPath, { encoding: "utf8" }));
  }

  verbose(`Writing css to ${cssOutputPath}\n`, css);

  try {
    writeCSS(sourceFilename, fullOutputPath, css);
  } catch (e) {
    error("Writing css failed", e);
  }

  return {
    propsToClassNamesMap: propKeyComboCSS.reduce((map: StringObject, [key, { selector }]) => {
      map[key] = selector;
      return map;
    }, {}),
    relevantPropKeys: propKeyComboCSS.reduce((set, [key, { keyCombo }]) => {
      return [...new Set([...set, ...keyCombo])];
    }, [] as string[]),
  };
}
