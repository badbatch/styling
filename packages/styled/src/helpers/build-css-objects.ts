import { buildMapKeyFromPropKeyCombo } from "@styling/helpers";
import { CSSVariablePropList, Interpolation, Metadata, StylingConfig } from "@styling/types";
import { PropKeyComboCSS } from "../types";
import buildBaseSelector from "./build-base-selector";
import buildInterpolationProps from "./build-interpolation-props";
import buildSelectorFromPropKeyCombo from "./build-selector-from-prop-key-combo";
import collateCSS from "./collate-css";
import dedupeCSS from "./dedupe-css";
import orderPropKeyComboCSS from "./order-prop-key-combo-css";

export default function buildCSSObjects(
  propKeyCombos: string[][],
  cssVariablePropList: CSSVariablePropList,
  interpolations: Interpolation[],
  { componentName, propList, sourceFilename }: Metadata,
  { selectorPrefix, theme }: StylingConfig,
) {
  const baseSelector = buildBaseSelector(sourceFilename, componentName, selectorPrefix);
  const baseCSS = collateCSS(interpolations, buildInterpolationProps([], cssVariablePropList), theme);

  const propKeyComboCSS: PropKeyComboCSS = [];

  propKeyComboCSS.push([
    "base",
    {
      css: baseCSS,
      keyCombo: [],
      selector: baseSelector,
    },
  ]);

  const sortedKeyCombos = propKeyCombos.sort((a, b) => {
    return a.length - b.length;
  });

  sortedKeyCombos.forEach(keyCombo => {
    const comboCSS = dedupeCSS(
      collateCSS(interpolations, buildInterpolationProps(keyCombo, cssVariablePropList), theme),
      keyCombo,
      propKeyComboCSS,
      propList,
    );

    if (Object.keys(comboCSS).length) {
      propKeyComboCSS.push([
        buildMapKeyFromPropKeyCombo(keyCombo, propList),
        {
          css: comboCSS,
          keyCombo,
          selector: buildSelectorFromPropKeyCombo(baseSelector, keyCombo, propList),
        },
      ]);
    }
  });

  return orderPropKeyComboCSS(propKeyComboCSS, propList);
}
