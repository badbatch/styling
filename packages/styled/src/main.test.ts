import { getInterpolations } from "./__tests__/helpers";
import styled from "./main";

jest.mock("./helpers/write-css");
const writeCSS = require("./helpers/write-css") as { default: jest.Mock };

describe("styled", () => {
  const checkedInterpolation = getInterpolations`
    color: ${p => (p.disabled ? "grey" : "blue")};
  `;

  const { propsToClassNamesMap } = styled("div", ["disabled", "checked"], "Container", "path/to/component/index.ts")`
    font-family: Arial;
    font-size: 16px;
    line-height: 1.25;
    ${p => p.checked && checkedInterpolation};
    ${p => p.disabled && "curser: default"};
  `;

  it("SHOULD generate the correct props to class names map", () => {
    Object.keys(propsToClassNamesMap).forEach(key => {
      if (key === "base") return;

      expect(propsToClassNamesMap[key].startsWith("container")).toBe(true);
      expect(propsToClassNamesMap[key].endsWith(key)).toBe(true);
    });
  });

  it("SHOULD write the correct css to file", () => {
    expect(writeCSS.default.mock.calls[0][2]).toMatchSnapshot();
  });
});
