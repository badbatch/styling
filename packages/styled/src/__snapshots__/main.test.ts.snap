// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled SHOULD write the correct css to file 1`] = `
".container-596050158 {
  font-family: Arial;
  font-size: 16px;
  line-height: 1.25; }

.container-596050158--checked {
  color: blue; }

.container-596050158--disabled {
  curser: default; }

.container-596050158--disabled--checked {
  color: grey; }

"
`;
